data_structures/queue/priority_queue.py

```diff
@@ -0,0 +1,273 @@
+"""
+Pure Python implementation of Priority Queue using lists
+"""
+
+
+class Error(Exception):
+    """Base class for exceptions in this module."""
+
+    pass
+
+
+class OverFlow(Error):
+    def __init__(self, msg):
+        self.msg = msg
+
+
+class UnderFlow(Error):
+    def __init__(self, msg):
+        self.msg = msg
+
+
+class FixedPriorityQueue:
+    """
+    In a Priority Queue the elements are entred as an when the come
+    But while removing or deleting an element the highest priority element is deleted
+    in FIFO fashion.
+    Here the lowest integer has the highest priority.
+    Example:
+    priority(0) > priority(5)
+    priority(16) > priority(32)
+    Here as an example I have taken only 3 priorities viz. 0, 1, 2
+    0 Being the highest priority and 2 being the lowest
+    You can change the priorities as per your need
+
+    Examples
+    >>> FPQ = FixedPriorityQueue()
+    >>> FPQ.enqueue(0, 10)
+    >>> FPQ.enqueue(1, 70)
+    >>> FPQ.enqueue(0, 100)
+    >>> FPQ.enqueue(2, 1)
+    >>> FPQ.enqueue(2, 5)
+    >>> FPQ.enqueue(1, 7)
+    >>> FPQ.enqueue(2, 4)
+    >>> FPQ.enqueue(1, 64)
+    >>> FPQ.enqueue(0, 128)
+    >>> FPQ.print_queue()
+    Priority 0: [10, 100, 128]
+    Priority 1: [70, 7, 64]
+    Priority 2: [1, 5, 4]
+    >>> FPQ.dequeue()
+    10
+    >>> FPQ.dequeue()
+    100
+    >>> FPQ.dequeue()
+    128
+    >>> FPQ.dequeue()
+    70
+    >>> FPQ.dequeue()
+    7
+    >>> FPQ.print_queue()
+    Priority 0: []
+    Priority 1: [64]
+    Priority 2: [1, 5, 4]
+    >>> FPQ.dequeue()
+    64
+    >>> FPQ.dequeue()
+    1
+    >>> FPQ.dequeue()
+    5
+    >>> FPQ.dequeue()
+    4
+    >>> FPQ.dequeue()
+    Traceback (most recent call last):
+                    ...
+    priority_queue.UnderFlow: Under Flow!
+    """
+
+    def __init__(self):
+        self.queue = [
+            [],
+            [],
+            [],
+        ]
+
+    def enqueue(self, priority, data):
+        """
+        This function enters the element into the queue based on its priority
+        If the priority is invalid an Exception is raised saying Invalid Priority!
+        If the queue is full an Exception is raised saying Over Flow!
+        """
+        if priority > 2:
+            raise Exception("Invalid Priority!")
+        elif len(self.queue[priority]) == 100:
+            raise OverFlow("Over Flow!")
+        else:
+            self.queue[priority].append(data)
+
+    def dequeue(self):
+        """
+        Return the highest priority element in FIFO order.
+        If the queue is empty then an under flow exception is raised.
+        """
+        if not self.queue[0] and not self.queue[1] and not self.queue[2]:
+            raise UnderFlow("Under Flow!")
+
+        if len(self.queue[0]) != 0:
+            return self.queue[0].pop(0)
```

Comment:
`.pop(0)` will be an `O(n)` operation. Not that this is of particular concern here since we're hard limited to no more than 100 values. However, if with an arbitrary number of queues and an arbitrary number of items, the proposed algorithm would have `O(n)` runtime for extracting the maximum value. FIFO queues in python should use the [collections.deque](https://docs.python.org/3.8/library/collections.html#collections.deque) data structure which has append, appendleft, pop, and popleft operations.

Reply:
Here I am implementing the data structure using list so everything is kinda list oriented